### ISSUE TITLE
Question: is this possible to iterate k-mers stored in SBWT representation

### ISSUE THREAD

**Issue by marekkokot** [algbio/SBWT#6]
Hi,

great software, congrats!
Is it possible to list all k-mers in SBWT instead of querying a specific k-mer?
If so, how to do this, is this somehow exposed in the C++ API?

Best
Marek

**Comment by jnalanko** [comment]
Hi Marek,

Thank you for maintaining KMC! It's been very useful for us.

Retrieving the k-mer at index i in the SBWT would require select queries on the SBWT subset sequence. In the bit matrix representation (=the default), this is just a bit vector select query, so it would be easy, but it does require building a select support structure, which takes some extra space. In other representations, such as Split, it's not so straightforward. But we can always fall back to solving a select query by binary searching with log(n) rank-queries.

So, while this is currently not supported in the API, we could add a function that does this in O(k log n) time per k-mer, and we could add an optional support structure that would reduce this to O(k) time, at least for the bit matrix representation.

On the other hand if you want to just dump the k-mers out of the data structure all at once, this can be done very fast in cache-efficient O(nk) time without a select structure. The algorithm reconstructs the k-mers "column by column" so that in the first iteration, it reports the last character of every k-mer, then on the next iteration the second to last character... and so on. We implemented this algorithm for [this paper](https://arxiv.org/abs/2306.04850), but it's not exposed in the API currently. Would this kind of a k-mer dump work for you, or do you need to for example stream them one by one without having all of them in memory at once?

**Comment by jnalanko** [comment]
Actually, the O(nk) full k-mer dump is already in the API (I forgot I had added it earlier):

https://github.com/algbio/SBWT/blob/da4f90e97bdb2bad661f4238f99c0cd97db20c02/include/sbwt/SBWT.hh#L292

**Comment by jnalanko** [comment]
I added two new functions to the API to access the i-th k-mer in colexicographic order:

* `get_kmer`, which works for any subset rank structure, but takes O(k log n) time. For the bit matrix SBWT of 3682 E. coli genomes at k = 31, this takes about 30us / kmer on my machine. https://github.com/algbio/SBWT/blob/ebe43f4cb570c2dff0817633fbaf41d97a01762f/include/sbwt/SBWT.hh#L303
* `get_kmer_fast`, which is accelerated by a subset select data structure that is passed into the function. There is a new class `SubsetMatrixSelectSupport` which adds select support to the bit matrix SBWT, with a space overhead of about 20%. This speeds up the query to about 4 us / kmer on my machine. Other representations than the bit matrix are not supported yet. https://github.com/algbio/SBWT/blob/ebe43f4cb570c2dff0817633fbaf41d97a01762f/include/sbwt/SBWT.hh#L315

The select support can be built and used like this:

```c++
plain_matrix_sbwt_t sbwt;
sbwt.load(...);
SubsetMatrixSelectSupport<sdsl::bit_vector> ss(sbwt.get_subset_rank_structure());
vector<char> buf(k+1); // The k-mer will be written here
sbwt.get_kmer_fast(i, buf.data(), ss); // Get the k-mer with colex rank i
```

Note that the k-mer at colexicographic rank i may be a technical "dummy k-mer" that is prefixed by some number of dollar symbols. You can detect if a k-mer is a dummy k-mer by checking if the first character is '$'.

Here is an example for streaming over all non-dummy k-mers of a bit matrix SBWT using a select support: https://github.com/algbio/SBWT/blob/e7accc72deeb740e737ed801e3c4f84d6002bc78/api_examples/api_example.cpp#L43

**Comment by marekkokot** [comment]
Hi Jarno,

Thank you for your in-depth explanation!
I missed that `reconstruct_all_kmers` method was added (I was looking at fork of SBWT I made some time ago and forgot to check what new was added in the original repo) - sorry.
Thank you also for quickly implementing `get_kmer` and `get_kmer_fast` (and providing running times)! I think it is a huge advantage over `reconstruct_all_kmers` if the user would like to dump all k-mer strings into a text file (something like `dump` operation of `kmc_tools`) because of much lower memory usage.
I was just wondering what is possible with SBWT :)
I probably should understand all the algorithmic details of SBWT described in the paper - sometimes it is easier to just ask directly :)

Yet another question. Let's say we are iterating k-mers of SBWT using `get_kmer`, and we want to apply some filtering criteria. Is it possible to construct a new SBWT while iterating the existing one?
So basically, is something more or less like the pseudocode below possible?
```
SBWT new_sbwt

for co_lext_kmer in original_sbwt:
  if not filtered_out(co_lext_kmer):
    new_sbwt.add(co_lext_kmer)
```
or it's not that simple?

**Comment by jnalanko** [comment]
No, problem, happy to help!

Filtering an SBWT: It's not quite that simple because of the dummy nodes required. That is, for every k-mer x in the filtered data structure that does not have a predecessor in the de Bruijn graph (i.e. there does not exist a k-mer y such that y[2..k] = x[1..k-1]), we need to add all dollar-padded proper prefixes of x to the data structure. So for example if x = ACCG, we would need to add $$$$, $$$A, $$AC and $ACC. And we need to sort these dummy k-mers with the rest of the k-mers. It's not so simple because we don't know which dummy k-mers we need before we are done with the filtering. It would probably have to be done in three stages: 1) detect which k-mers are left without a predecessor 2) sort the distinct prefixes of those (using a trie, maybe?) 3) merge the sorted prefixes with the sorted filtered k-mers and write out the filtered SBWT.

**Comment by marekkokot** [comment]
I see. Thank you again for the explanation and for extending the API to iterate k-mers.
